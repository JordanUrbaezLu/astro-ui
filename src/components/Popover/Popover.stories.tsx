import { Story } from "@storybook/react";
import { Popover, PopoverProps } from "./Popover";
import { Button } from "../Button/Button";

export default {
  component: Popover,
  title: "Components/Popover",
};

const Template: Story<
  Omit<PopoverProps, "children" | "startsOpen" | "trigger">
> = (args) => {
  return (
    <Popover
      {...args}
      startsOpen
      trigger={<Button variant="secondary">Trigger</Button>}
    />
  );
};

export const Default = Template.bind({});
Default.args = {
  text: "Popover",
};

export const TopPosition = Template.bind({});
TopPosition.args = {
  text: "Popover",
  position: "top",
};

export const BottomPosition = Template.bind({});
BottomPosition.args = {
  text: "Popover",
  position: "bottom",
};

export const TopPositionAndNubbin = Template.bind({});
TopPositionAndNubbin.args = {
  text: "Popover",
  position: "top",
  hasNubbin: true,
};

export const BottomPositionAndNubbin = Template.bind({});
BottomPositionAndNubbin.args = {
  text: "Popover",
  position: "bottom",
  hasNubbin: true,
};

export const LongText = Template.bind({});
LongText.args = {
  text: "I am a Popover with very long text!",
  position: "bottom",
};

export const AllProps = Template.bind({});
AllProps.args = {
  text: "I am a Popover with very long text!",
  position: "top",
  hasNubbin: true,
};
