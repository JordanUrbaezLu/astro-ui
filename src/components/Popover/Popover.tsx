import * as React from "react";
import { CSSTransition } from "react-transition-group";
import styles from "./Popover.module.scss";
import classNames from "classnames";
import { getPopoverPositionCalculations } from "../../utility/getPopoverPositionCalculations";

export type PopoverPosition = "bottom" | "top";

export interface PopoverProps {
  /**
   * If the Popover has a nubbin
   *
   * @default false
   */
  hasNubbin?: boolean;
  /**
   * If the Popover is open
   */
  isOpen: boolean;
  /**
   * The position of the Popover
   *
   * @default "bottom"
   */
  position?: PopoverPosition;
  /**
   * The text for the Popover
   */
  text: string;
  /**
   * The callback fired when the Popover toggles
   */
  toggleOpen: () => void;
  /**
   * The trigger for the Popover
   */
  trigger: React.ReactElement;
}

export const Popover: React.FC<PopoverProps> = (props) => {
  const {
    hasNubbin = false,
    isOpen,
    position = "bottom",
    text,
    toggleOpen,
    trigger,
  } = props;

  const popoverRef = React.useRef<HTMLDivElement>(null);

  const [popoverPositionStyle, setPopoverPositionStyle] =
    React.useState<React.CSSProperties | undefined>(undefined);

  const triggerRef = React.useRef<HTMLDivElement>(null);

  const popover = classNames(
    styles.popover,
    position === "bottom" && styles.bottom,
    position === "top" && styles.top,
    hasNubbin && styles.nubbin
  );

  React.useLayoutEffect(() => {
    setPopoverPositionStyle(
      getPopoverPositionCalculations({
        popoverRef,
        triggerRef,
        position,
        hasNubbin,
      })
    );
  }, []);

  return (
    <div className={styles.container}>
      {React.cloneElement(trigger, {
        onClick: () => toggleOpen(),
        ref: triggerRef,
        role: "button",
        tabIndex: 0,
      })}
      <div
        className={styles.popoverLayoutContainer}
        style={popoverPositionStyle}
        ref={popoverRef}
      >
        <CSSTransition
          in={isOpen}
          timeout={250}
          mountOnEnter
          unmountOnExit
          classNames={
            position === "top"
              ? {
                  enter: styles.enterTop,
                  enterActive: styles.enterTopActive,
                  exit: styles.exitTop,
                  exitActive: styles.exitTopActive,
                }
              : {
                  enter: styles.enterBottom,
                  enterActive: styles.enterBottomActive,
                  exit: styles.exitBottom,
                  exitActive: styles.exitBottomActive,
                }
          }
        >
          <div className={popover}>{text}</div>
        </CSSTransition>
      </div>
    </div>
  );
};
