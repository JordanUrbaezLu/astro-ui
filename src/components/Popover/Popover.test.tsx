import { render, screen } from "@testing-library/react";
import { Popover, PopoverPosition } from "./Popover";
import { axe } from "jest-axe";

describe("Popover", () => {
  test("Should render component correctly.", () => {
    const { container } = render(
      <Popover
        isOpen
        text="Popover"
        toggleOpen={jest.fn}
        trigger={<button>Button</button>}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("Accessibility", () => {
  test("Should have no accessibility violations.", async () => {
    const { container } = render(
      <Popover
        isOpen
        text="Popover"
        toggleOpen={jest.fn}
        trigger={<button>Button</button>}
      />
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});

describe("Props", () => {
  test("Should render text correctly.", () => {
    render(
      <Popover
        isOpen
        text="Popover"
        toggleOpen={jest.fn}
        trigger={<button>Button</button>}
      />
    );
    expect(screen.getByText("Popover")).toBeInTheDocument();
  });
  test.each<PopoverPosition>(["top", "bottom"])(
    "Should render positions correctly.",
    (position) => {
      render(
        <Popover
          isOpen
          position={position}
          text="Popover"
          toggleOpen={jest.fn}
          trigger={<button>Button</button>}
        />
      );
      expect(screen.getByText("Popover")).toHaveClass(position);
    }
  );
  test("Should render trigger correctly.", () => {
    render(
      <Popover
        isOpen
        text="Popover"
        toggleOpen={jest.fn}
        trigger={<button>Button</button>}
      />
    );
    expect(screen.getByRole("button", { name: "Button" })).toBeInTheDocument();
  });
  test("Should render hasNubbin correctly.", () => {
    render(
      <Popover
        hasNubbin
        isOpen
        text="Popover"
        toggleOpen={jest.fn}
        trigger={<button>Button</button>}
      />
    );
    expect(screen.getByText("Popover")).toHaveClass("nubbin");
  });
  test("Should render isOpen correctly.", () => {
    render(
      <div data-testid="div">
        <Popover
          isOpen={false}
          text="Popover"
          toggleOpen={jest.fn}
          trigger={<button>Button</button>}
        />
      </div>
    );
    expect(screen.getByTestId("div")).not.toHaveTextContent("Popover");
  });
});
