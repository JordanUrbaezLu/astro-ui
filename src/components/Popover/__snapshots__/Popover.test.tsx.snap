// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover Should render component correctly. 1`] = `
<div
  class="container"
>
  <button
    role="button"
    tabindex="0"
  >
    Button
  </button>
  <div
    class="popoverLayoutContainer"
    style="transform: translate3d(0px, 5px, 0);"
  >
    <div
      class="popover bottom"
    >
      Popover
    </div>
  </div>
</div>
`;
